### DISCUSSION TITLE
Whichkey and telescope results both load very slow on startup

### DISCUSSION THREAD

**Discussion by neonvoidx** [LazyVim/LazyVim#4116] [Q&A]
https://github.com/user-attachments/assets/f1eae729-310c-43ad-a564-f6507d3e3657

I have `lazy=false` as well as `opts: { delay=50 }` but upon startup whichkey takes a while to load.

Same applies for telescope ( i also have it as lazy=false)

https://github.com/user-attachments/assets/1667f5dc-7081-4f28-ba8e-35901ccb1ab0

**Comment by dpetka2001** [comment]
I believe you also have to set `vim.opt.timeoutlen` to a low value. I'm honestly entirely not sure about the relationship between `opts.delay` and `vim.opt.timeoutlen`, but it seems that `which-key` takes both into consideration somehow and makes some kind of decision. Better to wait for an answer from the maintainer himself, since he's the most appropriate to give a proper explanation.

**Reply by neonvoidx** [reply]
mine is currently set to 300, makes no difference, feels like which key is just loading slow, and same thing with telescopes result (not telescope itself). I also set priority of which key and telescope to 900, no change

**Reply by dpetka2001** [reply]
Try setting `vim.opt.timeoutlen = 0` and `opts.delay = 50` and see how it behaves. No idea about Telescope, been using `fzf-lua` for a while now (although this also might depend on the size of the project you're searching in).

**Reply by folke** [reply]
That is not a good idea. Don't set `timeoutlen` to `0`.
Which-key no longer depends on `timeoutlen` in any way.

**Comment by folke** [comment]
I can't see those videos?

Also can't reproduce. You can even set `delay=0` to show which-key right away.
I also don't see how any LazyVim setting could impact showing slow telescope results.

**Reply by dpetka2001** [reply]
@folke Is there some other factor that affects the `opts.delay` setting? 

I'm observing the following. In the user configuration directory, when I press `<leader>` it shows faster. 
In the local forked LazyVim directory the delay is slightly bigger.

Here's a screecast, not sure how well the difference shows compared to real time

[Screencast 2024-07-20 14:03:48.webm](https://github.com/user-attachments/assets/10831961-4e3d-44ba-9a62-c73609264c8c)

I've also recorded the keys. If you see closely, in the user configuration directory the moment space is pressed (`󱁐` icon bottom right) the `which-key` menu shows up at the moment the icon appears (which corresponds to the time I'm pressing `<leader>`). In the forked LazyVim directory you can see the icon bottom right appear first and after a slight delay the `which-key` menu shows up. It doesn't show up at the exact moment `<leader>` is pressed like in the user configuration directory.

**Reply by folke** [reply]
You probably didnt update which-key in one of the two?

**Reply by dpetka2001** [reply]
It's the same instance of Neovim. I've called it LazyVimDev, because it corresponds to the locally forked LazyVim repo. In both cases when I launch Neovim I choose `LazyVimDev` from the menu (the only difference is I'm launching 2 instances of Neovim with the same profile in 2 different directories). `which-key` is updated to latest version.

PS: It's not a big deal really. It just seemed awkward to me this different behavior in 2 different directories.

**Reply by folke** [reply]
Sorry, I find these screenscasts really hard to follow.
But as far as I understand, you do the exact same thing in two different directories.
First time in `~/.config/nvim` and then in `~/projects/LazyVim`.


---

While writing this comment, I thought of something. Will check

**Reply by folke** [reply]
Can you try again, but without using a session?

**Reply by folke** [reply]
And maybe set `delay = 1000` or something, so that it's easier for me to see what's happening.

**Reply by dpetka2001** [reply]
Well you were right about the session. Without restoring a session and just opening Neovim with a file argument both Neovim instances in the 2 different directories (same LazyVimDev profile) show the `which-key` menu now the same fast. If there's a difference it's not discernible to my eyes at least. I don't think I need to make a new screencast with `delay = 1000`? Or would you still want me to?

**Reply by folke** [reply]
Hmmm, odd. I don't really get how that may yield differerent results.

Did you manually change your `sessionoptions` to include `localoptions` or `options`?
That also saves buffer local keymaps, but the descriptions would not be saved.
This way, which-key wouldn't know if a keymap is one created by which-key.

Or can you check `:set ttl?` in both sessions? Maybe a differernt value was saved in the session?

**Reply by dpetka2001** [reply]
No, haven't overwritten anything of these. Below is what `verbose` shows
```lua
verbose set sessionoptions?   sessionoptions=buffers,curdir,tabpages,winsize,help,globals,skiprtp,folds
	Last set from ~/projects/plugins/LazyVim/lua/lazyvim/config/options.lua line 89
```
Also by `:set ttl?`, did you actually mean `:set ttm?` maybe? On both sessions the value of that is `50`. Also I deleted all files under `~/.local/state/LazyVimDev/sessions` to start fresh and this still happens. There's a small but discernible difference to my eyes, when restoring a session in the 2 different directories (even with fresh new sessions after deleting). But when opening with file argument in the 2 different directories the amount of time for which-key menu to show up looks to me the same. 

Anyway, this seems kind of nitpicking from my part, so don't really want to bother you further. Like I said it's really not a big deal (not to the point that it bothers me), so all good.

**Comment by neonvoidx** [comment]
I tried this on my personal laptop and seems fine, I think it is honestly just crowdstrike on my work laptop bringing things to a slow halt during scans

**Comment by mbfoss** [comment]
I am not using Lazy at all and still encountered this issue with which-key.
It seems the trick is that which-key should be loaded LAST (maybe because other plugins already mapped their keys?)
At least this is what improved loading speed for me.
